### DISCUSSION TITLE
Passing Ref to Providers / Services

### DISCUSSION THREAD

**Discussion by LucaCras** [rrousselGit/riverpod#1188] [Q&A] [answered]
I am a bit unclear on some documentation and what is the correct way of handling the (Widget)Ref objects in certain cases and whether I should worry about memory leaks of some sort:
1. Is it okay to pass a ref to a service class in its constructor without inadvertently leaking memory? Like this:
`final chartInteractionServiceProvider = Provider.family<ChartInteractionService, String>((ref, id) => ChartInteractionService(ref, id));`
2. Is it better / worse to pass a WidgetRef as a parameter to a method of said service class? e.g.:
`ref.read(chartInteractionServiceProvider(simulationId)).onTouchMove(ref, touchPosition);`
3. I am working with behaviour that, depending on some state, either requires updating some other state or not. This is why I want to pass the ref, so that it only gets the stateControllers in those cases where it is needed. Would you say this is alright, or that this necessity is a code smell or a sign of bad structure/architecture?
4. In the docs, it is stated that read should/can be passed to a provider / service, but I have also read that it is preferrable to use
`ref.watch(X.notifier)` compared to `ref.read(X)` what should I make of this in the case of a method in a service that only updates state (without the need to watch / read the actual state) and what if it does need to read the state before updating the state?

Thank you in advance!

**Comment by rrousselGit** [comment, marked as answer]
Making Ref to objects is fine.

Passing WidgetRef is not. Refactor code that depends on WidgetRef to use a Provider's Ref instead.

3) by itself, there's nothing wrong with that. Of course it depends on what exactly you're doing

4) within a method for triggering an update, use read.

**Reply by LucaCras** [reply]
This is the method in question:

```
void onTouchMove(WidgetRef ref, Offset currentTouchPosition, DrugMode mode) {
    final chartState = ref.read(chartStateProvider(simulationId));

    switch (chartInteraction) {
      case ChartInteractionType.MoveXValue:
        chartState.dx = currentTouchPosition.dx;
        break;

      case ChartInteractionType.MoveYValue:
        chartState.dy = currentTouchPosition.dy;
        break;

      case ChartInteractionType.ZoomYLeftAxis:
        final currentZoom = ref.read(leftYAxisZoomProvider(simulationId));
        _zoomYAxis(chartState, currentZoom, currentTouchPosition);
        break;

      case ChartInteractionType.ZoomYRightAxis:
        final currentZoom = ref.read(rightYAxisZoomProvider(simulationId));
        _zoomYAxis(chartState, currentZoom, currentTouchPosition);
        break;

      case ChartInteractionType.ZoomPanXAxis:
        final currentZoom = ref.read(xAxisZoomProvider);
        _zoomPanXAxis(chartState, currentZoom, currentTouchPosition);
        break;

      default:
        return;
    }
    switch (mode) {
      case DrugMode.MANUAL:
        final input = ref.read(inputProvider);
        _updateInput(chartState, input);
        break;
      case DrugMode.TCI:
      case DrugMode.ETCI:
        final target = ref.read(targetProvider);
        _updateTarget(chartState, target);
        break;
      case DrugMode.IVASSIST:
        final ivAssistTarget = ref.read(ivAssistTargetProvider);
        _updateIVAssistTarget(chartState, ivAssistTarget);
        break;
    }
  }
```
